README.md
Ouroboros
==============
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1452511.svg)](https://doi.org/10.5281/zenodo.1452511)

[![Build Status](https://travis-ci.org/miguelcorrea/Ouroboros.svg?branch=master)](https://travis-ci.org/miguelcorrea/Ouroboros)

This is the repository for Ouroboros, an algorithm for intermolecular coevolutionary analysis. Ouroboros is meant to be used when the multiple sequence alignments to be analyzed contain pairs of non-interacting sequences, which decreases intermolecular contact prediction performance. Ouroboros attempts to identify, without any prior information, which pairs are interacting and which are not. These protein-protein interaction predictions can in turn be used to improve contact prediction performance.

# FAQ

## How does it work?
Ouroboros combines coevolutionary analysis with expectation-maximization to simultaneously model protein-protein interaction and intermolecular contacts. It alternates between two steps: predicting protein-protein interaction based on the model describing intermolecular contacts, and estimating the model parameters based on protein-protein interaction predictions. By weighting proteins in the models according to the protein-protein interaction predictions, we boost the coevolutionary signal. For further details, please [read our paper](https://doi.org/10.1093/bioinformatics/bty924).

## How do I run it?
The *run_analysis.py* script takes only one argument, the path to a JSON file that contains options for the arguments. Some examples of these can be found in data/params.
It is required that the file contains at least:
* io: path where results will be saved
* msa1: path to the first MSA
* msa2: path to the second MSA
* init: how to initialize the model: either "warm" or "random"
* mode: whether to perform "soft" or "hard" EM
* int_frac: prior fraction of interacting proteins (between 0 and 1)

If you know the ground truth about your input and want to test performance, you will also need to pass:
* test: set to True
* int_limit: index of the last interacting sequence pair in your dataset. Please prepare the MSAs so that the interacting sequences are at the top, and the non-interacting ones at the bottom. 
* contact_mtx: (optional) path to the contact matrix in CSV format. There should be as many rows as columns in msa1, and as many columns as in msa2.

As additional options, you can set:
* n_jobs: number of CPUs to use in model fitting (default: 2)
* n_starts: number of random starts to perform (if using a "random" start, ignored otherwise; default: 5)
* dfmax: maximum number of degrees of freedom allowed in model fitting (default:100)
* gap_threshold: gap frequency threshold; MSA columns with a gap frequency above this threshold are deleted (default: 0.5)
* tol: tolerance of the algorithm in soft EM, below which we consider the algorithm converges (default:0.005)
* max_iters: maximum number of EM iterations to perform (default:20)
* max_init_iters: maximum number of logistic regression iterations during model initialization (default: 100)
* max_reg_iters: maximum number of logistic regression iterations after initialization (default: 750)
* predict_contacts: boolean, whether to perform a final round of contact prediction or not. Although we have not focused in this area, and it is not extensively tested, it is possible obtain contact predictions.

## How do I interpret the output?
The most important files to look at in the results are:
* labels_per_iter.csv is a matrix (of dimensions sequence pairs x EM iterations) containing the interaction probability for each sequence pairs over iterations of EM
* output/z_over_iters.pdf shows the evolution of the protein-protein interaction probabilities over EM iterations
* output/convergence.png shows the evolution of the log-likelihoods

If you have passed ground truth about the interactions, you will obtain some additional files:
* true_alt_llhs.csv contains the log-likelihood of the ground truth according to the coevolutionary model over EM iterations
* true_null_llhs.csv contains the log-likelihood of the ground truth according to the independent model over EM iterations
* true_total_llhs.csv contains the total log-likelihood of the ground truth over EM iterations
* output/model_report.txt contains the model performance (as cross-entropy log-loss and Matthews Correlation Coefficient).
* output/conf_mtx.png and norm_conf_mtx.png contain confusion matrices; the second one is normalized to take into account class imbalance.
* output/perf_per_iter.png shows the evolution of the performance over EM iterations

Some additional files provide more detailed information:
* all_alt_llhs.csv contains the log-likelihood of the data according to the coevolutionary model over EM iterations
* all_null_llhs.csv contains the log-likelihood of the data according to the the independent evolution model over EM iterations
* all_total_llhs.csv contains the total log-likelihood of the data over EM iterations
* num_mtx_*.csv contains the alignments as numeric matrices  
* bin_mtx_*.csv contains the one-hot encoded alignments (binary matrices in the paper)
* processed_contact_matrix.csv (if you have passed a contact matrix) contains the ground truth contact matrix once it has been processed like the input alignments (i.e. removal of constant and gappy columns).  
* output/alt_llhs_mtx_*.csv contains the log probability of each individual residue at the concatenated alignments at a particular iteration according to the coevolutionary model
* output/null_llhs_mtx_*.csv contains the log probability of each individual residue at the concatenated alignments at a particular iteration according to the null model
* output/fixed_alphas_*_iter_init.csv contains the selected regularization strengths for each column for MSA A and B.
* output/dfs_*_iter_init.csv contains the degrees of freedom of the models selected during initialization

If you choose to perform contact prediction, you will obtain some additional output:
* output/final_contact_mtx.csv is a matrix (of dimensions msa1 x msa2, once they have been processed) containing coevolutionary strengths between all pairs of residues between your two MSAs
* output/norm_final_contact_mtx.csv is the same matrix, normalized using the Average Product Correction of Dunn *et al.* (2008)

Other undocumented output is present for development and testing reasons, and might be removed in the future.

## A warning tells me the log-likelihood is not increasing monotonically.
In hard EM, the log-likelihood is not mathematically guaranteed to increase monotonically, and so this does not necessarily indicate a problem. In soft EM, however, it is theoretically supposed to. We have observed that, in rare cases, and when the log-likelihood of the data has plateaued, one of the logistic regression models does not estimate the parameters well enough, leading to small decreases in log-likelihood. Although we do not expect such small decreases to negatively affect the results, if this happens, you should inspect your results, especially the evolution of the log-likelihood. This issue can be solved by simply increasing the maximum number of iterations of logistic regression (i.e. the max_reg_iters parameter) in order to improve estimation of model coefficients.

## How do I cite it?
If you use our software in your research, please cite our paper:
Correa Marrero et al. "Improved inference of intermolecular contacts through protein-protein interaction prediction using coevolutionary analysis." Bioinformatics (2018).

# Contributing
We welcome contributions to improve our software, including feature additions, unit tests, bug fixes and improvements to documentation. 

# License
This is an open source tool available under the BSD-3 license. See the LICENSE file for details.

# Important links
* Source code repo: https://github.com/miguelcorrea/Ouroboros
* Issue tracker: https://github.com/miguelcorrea/Ouroboros/issues
* Paper: https://doi.org/10.1093/bioinformatics/bty924

# Requirements
* Python 3.6+
* NumPy
* scikit-learn
* scikit-bio
* matplotlib
* seaborn
* tqdm

CHANGELOG.md
# CHANGELOG

### v 0.3 (June 2019)
---
* Optimized model fitting code. This leads to a drastic decrease in running time without sacrificing model performance. An example case with the PDZ data is now around 18 times faster, while an example with the TCS data runs 27 times faster.
* Now requires at least scikit-learn v. 0.20
* Removed unused code

### v 0.2 (October 2018)
---
* Added unit testing
* Added the *input_handling* module
* Improvements in how user-provided parameters are checked and validated
* Improved validation of input alignments
* The code now produces a README file explaining different output files
* Improvements to project README file
* Added two-component system example data and parameter file
* Removed prior_inters option
* Added predict_contacts option (experimental, default: False)

### v 0.1 (March 2018)
---
First release

code/src/input_handling.py
"""
Module containing functions to digest and validate input
"""
import os
import warnings
import json
import numpy as np


def validate_alignments(num_mtx_a, num_mtx_b):
    """
    Perform sanity checks on input alignments
    """
    assert not np.array_equal(num_mtx_a, num_mtx_b), """Alignments A and B
    are identical"""
    n_obs_a = num_mtx_a.shape[0]
    n_obs_b = num_mtx_b.shape[0]
    assert n_obs_a == n_obs_b, f"""Different number of sequences in MSA A
    ({n_obs_a}) and MSA B ({n_obs_b})"""
    assert n_obs_a > 1, "Only one sequence in the input alignments"


def validate_contact_mtx(msa_a, msa_b, contact_mtx):
    """
    Perform sanity check on input contact matrix
    """

    # Verify that the input contact matrix has the expected dimensions
    assert contact_mtx.shape == (msa_a.shape[1], msa_b.shape[1]), f"""Contact
    matrix is expected to have dimensions
    {msa_a.shape[1]} x {msa_b.shape[1]} (MSA A columns x MSA B columns);
    found {contact_mtx.shape} instead"""
    # Values in the contact matrix must be either 1 or 0
    assert set(np.unique(contact_mtx)) == {0, 1}, """ AAA """


def read_args(json_path):
    """
    Read JSON file containing analysis parameters
    """
    with open(json_path, 'r') as source:
        args = json.load(source)
    return args


def digest_args(args):
    """
    Process and validate input parameters
    """
    io_path = args['io']
    msa_a_path, msa_b_path = digest_msa_paths(args)

    gap_threshold = digest_gap_threshold(args)
    int_frac = digest_int_frac(args)
    init = digest_init(args)
    mode = digest_mode(args)
    test = digest_test(args)
    int_limit = digest_int_limit(args, test)

    contact_mtx = digest_contact_mtx(args)

    n_jobs = digest_n_jobs(args)
    n_starts = digest_n_starts(args, init)
    dfmax = digest_dfmax(args)

    max_init_iters = digest_max_init_iters(args)
    max_reg_iters = digest_max_reg_iters(args)
    predict_contacts = digest_pred_contacts(args)

    return io_path, msa_a_path, msa_b_path, gap_threshold, int_frac, init, \
        mode, test, int_limit, contact_mtx, n_jobs, n_starts, dfmax, max_init_iters,\
        max_reg_iters, predict_contacts


def digest_msa_paths(args):
    if 'msa1' in args.keys():
        msa_a_path = args['msa1']
    else:
        raise KeyError('Missing mandatory parameter: msa1')
    if 'msa2' in args.keys():
        msa_b_path = args['msa2']
    else:
        raise KeyError('Missing mandatory parameter: msa2')

    if not os.path.isfile(msa_a_path):
        raise ValueError('Path to MSA A is not a file')
    if not os.path.isfile(msa_b_path):
        raise ValueError('Path to MSA B is not a file')
    if msa_a_path == msa_b_path:
        raise ValueError('Path to MSA A and MSA B are the same')
    return msa_a_path, msa_b_path


def digest_int_frac(args):
    if 'int_frac' in args.keys():
        int_frac = args['int_frac']
    else:
        raise KeyError('Missing mandatory parameter: int_frac')

    if int_frac <= 0 or int_frac >= 1:
        raise ValueError(f'Invalid value of int_frac {int_frac}')
    return int_frac


def digest_init(args):
    if 'init' in args.keys():
        init = args['init']
    else:
        raise KeyError('Missing mandatory parameter: init')

    if init != "warm" and init != "random":
        raise ValueError(f"""Invalid value of init: {init}.
            Only warm and random accepted""")
    return init


def digest_mode(args):
    if 'mode' in args.keys():
        mode = args['mode']
    else:
        raise KeyError('Missing mandatory parameter: mode')

    if mode != "hard" and mode != "soft":
        raise ValueError(f"""Invalid value of mode: {mode}.
            Only hard and soft accepted""")
    if mode == "hard":
        warnings.warn("""Chosen hard EM. This option is at the moment less
            extensively tested than soft EM.""", UserWarning)
    return mode


def digest_test(args):
    if 'test' in args.keys():
        if type(args['test']) == bool:
            test = args['test']
        else:
            raise ValueError(f"Invalid, non-boolean value for test parameter: {args['test']}")
    else:
        test = False
    return test


def digest_gap_threshold(args, default=0.5):
    if 'gap_threshold' in args.keys():
        gap_threshold = args['gap_threshold']
        if gap_threshold >= 1 or gap_threshold <= 0:
            raise ValueError(f"""Gap threshold value {gap_threshold} outside
                bounds""")
    else:
        gap_threshold = default
    return gap_threshold


def digest_int_limit(args, test):
    if 'int_limit' in args.keys() and test is True:
        int_limit = args['int_limit']
        if int_limit < 0:
            raise ValueError(f'Invalid value of int_limit: {int_limit}')
    elif test is True:
        raise ValueError(
            'int_limit argument is mandatory when using test mode')
    elif 'int_limit' in args.keys():
        warnings.warn(
            'Passed value for int_limit without setting test to True; ignoring option')
        int_limit = None
    else:
        int_limit = None

    return int_limit


def digest_contact_mtx(args):
    if 'contact_mtx' in args.keys():
        if 'predict_contacts' in args.keys() and args['predict_contacts']:
            contact_mtx = args['contact_mtx']
            if not os.path.isfile(contact_mtx):
                raise ValueError('Path to contact matrix is not a file')
        else:
            warnings.warn("""Given a contact matrix, but not asked to predict
                contacts; ignoring contact matrix""", UserWarning)
            contact_mtx = None
    else:
        contact_mtx = None
    return contact_mtx


def digest_n_starts(args, init, default=5):
    if 'n_starts' in args.keys():
        if init == 'warm':
            n_starts = None
        elif init == 'random':
            n_starts = args['n_starts']
            if n_starts < 1:
                raise ValueError(f'Invalid value of n_starts: {n_starts}')
    elif init == 'random':
        n_starts = default
    else:  # Not applicable in warm start
        n_starts = None
    return n_starts


def digest_n_jobs(args, default=2):
    if 'n_jobs' in args.keys():
        n_jobs = args['n_jobs']
        if n_jobs < 1:
            raise ValueError(f'n_jobs value {n_jobs} outside bounds')
    else:
        n_jobs = default
    cpus = os.cpu_count()
    if n_jobs > cpus:
        warnings.warn(f"""n_jobs value {n_jobs} greater than available number
            of CPUs ({cpus}); setting value to {cpus}""")
        n_jobs = cpus
    return n_jobs


def digest_max_init_iters(args, default=100, recommend_min=50):
    if 'max_init_iters' in args.keys():
        max_init_iters = args['max_init_iters']
        if max_init_iters <= 1:
            raise ValueError(f"""Invalid value of max_init_iters:
                {max_init_iters}""")
        if max_init_iters < recommend_min:
            warnings.warn(f"""Low value ({max_init_iters}) of max_init_iters:
                may lead to improper model selection""")
    else:
        max_init_iters = default
    return max_init_iters


def digest_max_reg_iters(args, default=1000, recommend_min=750):
    if 'max_reg_iters' in args.keys():
        max_reg_iters = args['max_reg_iters']
        if max_reg_iters <= 1:
            raise ValueError(f"""Invalid value of max_reg_iters:
                {max_reg_iters}""")
        if max_reg_iters <= recommend_min:
            warnings.warn(f"""Low value ({max_reg_iters}) of max_reg_iters:
             may lead to optimization problems""")
    else:
        max_reg_iters = default
    return max_reg_iters


def digest_pred_contacts(args, default=False):
    if 'predict_contacts' in args.keys():
        if type(args['predict_contacts']) == bool:
            predict_contacts = args['predict_contacts']
        else:
            raise ValueError(f"""Invalid, non-boolean value for
                predict_contact parameter: {args['predict_contacts']}""")
    else:
        predict_contacts = default
    return predict_contacts

########################
# EM keyword arguments #
########################


def pack_em_kwargs(args, true_labels):
    """
    Prepare keyword arguments for EM loop
    """
    em_kwargs = {}

    em_kwargs['tol'] = digest_tol(args)
    em_kwargs['max_iters'] = digest_max_iters(args)
    em_kwargs['dfmax'] = digest_dfmax(args)
    em_kwargs['true_labels'] = true_labels

    return em_kwargs


def digest_tol(args, default=5e-3):
    if 'tol' in args.keys():
        tol = args['tol']
        if tol <= 0:
            raise ValueError(f"Invalid tolerance value: {tol}")
        machine_epsilon = np.finfo(float).eps
        if tol <= machine_epsilon:
            raise ValueError(f"Tolerance set below machine precision")
    else:
        tol = default
    return tol


def digest_max_iters(args, default=20):
    if 'max_iters' in args.keys():
        max_iters = args['max_iters']
        if max_iters <= 1:
            raise ValueError(f"Invalid max_iters value: {max_iters}")
    else:
        max_iters = default
    return max_iters


def digest_dfmax(args, default=100):
    if 'dfmax' in args.keys():
        dfmax = args['dfmax']
        if dfmax <= 1:
            raise ValueError(f'Invalid value of dfmax: {dfmax}')
        if dfmax >= 120:
            warnings.warn(f"""High setting of dfmax ({dfmax}).
                Allowing many degrees of freedom might lead to selection of
                overtly complex models.""")
    else:
        dfmax = default
    return dfmax
